scylla-cdc/src/consumer.rs

```diff
@@ -0,0 +1,194 @@
+                x => {
+                    if x.starts_with(IS_DELETED_PREFIX) {
+                        deleted_mapping.insert(x.to_string(), j);
+                    } else if x.starts_with(ARE_ELEMENTS_DELETED_PREFIX) {
+                        deleted_el_mapping.insert(x.to_string(), j);
+                    } else {
+                        mapping.insert(x.to_string(), j);
+                    }
+                    j += 1;
+                }
+            }
+        }
+
+        CDCRowSchema {
+            stream_id,
+            time,
+            batch_seq_no,
+            end_of_batch,
+            operation,
+            ttl,
+            mapping,
+            deleted_mapping,
+            deleted_el_mapping,
+        }
+    }
+}
+
+pub struct CDCRow {
+    pub stream_id: Vec<u8>,
+    pub time: uuid::Uuid,
+    pub batch_seq_no: i32,
+    pub end_of_batch: bool,
+    pub operation: OperationType,
+    // Can be NULL in the database.
+    pub ttl: Option<i64>,
+    data: Vec<Option<CqlValue>>,
+    // Maps element name to its index in the data vector.
+    mapping: HashMap<String, usize>,
+    deleted_mapping: HashMap<String, usize>,
+    deleted_el_mapping: HashMap<String, usize>,
+}
+
+impl CDCRow {
+    pub fn from_row(row: Row, schema: &CDCRowSchema) -> CDCRow {
+        let mapping = schema.mapping.clone();
+        let deleted_mapping = schema.deleted_mapping.clone();
+        let deleted_el_mapping = schema.deleted_el_mapping.clone();
+
+        // If cdc read was successful, these default values will not be used.
+        // [TODO]: Test it.
+        let (mut stream_id, mut time, mut batch_seq_no, mut end_of_batch, mut operation, mut ttl) = (
+            vec![],
+            uuid::Uuid::default(),
+            0,
+            true,
+            OperationType::PreImage,
+            None,
+        );
+        let mut data: Vec<Option<CqlValue>> = vec![];
+        let mut i = 0;
+
+        // .iter().enumerate() can't be used here, because it doesn't take ownership of taken value.
+        for column in row.columns.into_iter() {
+            // I have no idea why, but I couldn't use match here - variables were not accessible inside the match block.
+            if i == schema.stream_id {
+                stream_id = column.unwrap().into_blob().unwrap();
+            } else if i == schema.time {
+                time = column.unwrap().as_uuid().unwrap();
+            } else if i == schema.batch_seq_no {
+                batch_seq_no = column.unwrap().as_int().unwrap();
+            } else if i == schema.end_of_batch {
+                end_of_batch = column.unwrap().as_boolean().unwrap()
+            } else if i == schema.operation {
+                operation = OperationType::try_from(column.unwrap().as_tinyint().unwrap()).unwrap();
+            } else if i == schema.ttl {
+                ttl = column.map(|ttl| ttl.as_bigint().unwrap());
+            } else {
+                data.push(column);
+            }
```

Comment:
I don't know why, but when I tried to do something like that:
`match i {
    schema.stream_id = > (...)
}`
it just didn't work - as if the variable `schema` was out of scope.